# Analyzing Data of Used Car Listings on eBay Kleinanzeigen

In this project, I will be working with a dataset of used cars from *eBay Kleinanzeigen*, a classifieds section of the German eBay website. The original dataset can be found [here](https://data.world/data-society/used-cars-data).

The data dictionary provided with the data is as follows:
* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
* `name` - Name of the car.
* `seller` - Whether the seller is private or a dealer.
* `offerType` - The type of listing
* `price` - The price on the ad to sell the car.
* `abtest` - Whether the listing is included in an A/B test.
* `vehicleType` - The vehicle Type.
* `yearOfRegistration` - The year in which the car was first registered.
* `gearbox` - The transmission type.
* `powerPS` - The power of the car in PS.
* `model` - The car model name.
* `kilometer` - How many kilometers the car has driven.
* `monthOfRegistration` - The month in which the car was first registered.
* `fuelType` - What type of fuel the car uses.
* `brand` - The brand of the car.
* `notRepairedDamage` - If the car has a damage which is not yet repaired.
* `dateCreated` - The date on which the eBay listing was created.
* `nrOfPictures` - The number of pictures in the ad.
* `postalCode` - The postal code for the location of the vehicle.
* `lastSeenOnline` - When the crawler saw this ad last online.

The goal of this project will be to clean & analyze the data of used car listings.

*Note: I will be simulating a less-cleaned version of a 50,000 data point sample from the above dataset.*

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv('autos.csv',encoding='Latin-1')
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The dataset contains 50,000 rows with 20 columns. All of the data is stored as either strings or integers, and 5 columns contain some number of null values.

## Modifying Column Labels
The current column labels are as follows:

In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

Below, I will modify the column labels to snakecase (for readability), and rename certain columns to better reflect the column descriptions as given by the data dictionary:

In [4]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

## Initial Data Exploration & Data Cleaning

To start, I will review the data for any obvious areas that require data cleaning.

In [5]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-10 15:36:24  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

Based on my initial observations, there are two columns in which nearly all (49999/50000) the values are the same:
* `seller`
* `offer_type`

Due to this fact, neither of these columns offer useful information for analysis, and can likely be dropped. There are also instances of numeric data being stored as strings:
* `price`
* `odometer`

These values will need to be converted appropriately. Furthermore, the values in the `num_photos` column indicate a need for more investigation.

In [6]:
#exploring all values in the num_photos column
autos['num_photos'].value_counts()

0    50000
Name: num_photos, dtype: int64

After further investigation of the `num_photos` column, one can see that every row holds a value of 0. Since all values of this column are the same, it can be dropped with the other two columns noted above.

In [7]:
autos = autos.drop(['seller','offer_type','num_photos'],axis=1)

Now, to clean and convert the numeric values in the `price` and `odometer` columns:

In [8]:
#first five rows of price &  odometer columns
autos[['price','odometer']].head()

price   odometer
0  $5,000  150,000km
1  $8,500  150,000km
2  $8,990   70,000km
3  $4,350   70,000km
4  $1,350  150,000km

In [9]:
#removing '$' and ',' characters from price, converting to integers
autos['price'] = (autos['price']
                  .str.replace('$','')
                  .str.replace(',','')
                  .astype(int)
                 )

#removing ',' and 'km' characters from odometer, converting to integers
autos['odometer'] = (autos['odometer']
                  .str.replace(',','')
                  .str.replace('km','')
                  .astype(int)
                 )
#renaming odometer to odometer_km
autos.rename({'odometer':'odometer_km'},axis=1,inplace=True)

#first 5 rows of new price & odometer_km columns
autos[['price','odometer_km']].head()

price  odometer_km
0   5000       150000
1   8500       150000
2   8990        70000
3   4350        70000
4   1350       150000

## Exploring Price & Odometer

In [10]:
print(autos["price"].describe())
print('\nNumber of Unique Values:')
print(autos["price"].unique().shape)
autos["price"].value_counts().head(10)

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

Number of Unique Values:
(2357,)


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
Name: price, dtype: int64

The `price` column has 2,357 unique values, with a minimum of \\$0. There are 1,421 listings with this minimum value, however, this is only about 2\% of all listings, so they can likely be dropped from our dataset.

This column also has a maximum value of \\$100,000,000, which seems like an excessive amount for a used car.

In [11]:
print('Most Expensive Listings:')
print(autos['price'].value_counts().sort_index(ascending=False).head(10))

print('\nPercent of Listings Valued Over $100,000:')
print((autos[autos['price']>100000].shape[0]/autos.shape[0])*100,'%')

Most Expensive Listings:
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64

Percent of Listings Valued Over $100,000:
0.106 %


There are a number of extremely expensive listings, with 14 of them being near or over \\$1 million dollars. However, the listings valued at over \\$100,000 comprise **less than a single percent** of the entire dataset. Setting a maximum value of \\$100,000 seems much more reasonable for online used car listings, so I will be dropping any higher listings as outliers.

In [12]:
#dropping rows with price < 1 and price > 100000
autos = autos[autos['price'].between(1,100000,inclusive='both')]
autos['price'].describe()

count    48526.000000
mean      5755.574146
std       7514.501772
min          1.000000
25%       1200.000000
50%       3000.000000
75%       7450.000000
max      99900.000000
Name: price, dtype: float64

Moving on to the `odometer_km` column:

In [13]:
print(autos['odometer_km'].describe())
print('\nOdometer Value Counts:')
autos['odometer_km'].value_counts().sort_index()

count     48526.000000
mean     125837.077031
std       39714.202576
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

Odometer Value Counts:


5000        827
10000       250
20000       754
30000       774
40000       814
50000      1010
60000      1153
70000      1217
80000      1414
90000      1734
100000     2113
125000     5055
150000    31411
Name: odometer_km, dtype: int64

The `odometer_km` column has only 13 unique values, all of which are rounded to the nearest 1,000 km, 10,000 km or 25,000 km. This likely implies that sellers were only given these specific values as options from a pre-set list. The majority of listings (roughly 65%) have an `odometer_km` value of 150,000 km, so the dataset is skewed toward high mileage vehicles.

## Converting & Exploring Date Columns

There are a number of columns with date information:
* `date_crawled`
* `registration_year`
* `registration_month`
* `ad_created`
* `last_seen`

In particular, the `date_crawled`, `ad_created`, & `last_seen` columns have dates being stored as strings, so these will need to be converted before continuing.

In [14]:
autos[['date_crawled','ad_created','last_seen']].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

Some of the above columns also include a time signature. Moving forward, I will only be using the dates for analysis, so I will also remove the times from the columns as well as convert them to datetime objects.

In [15]:
import datetime as dt

#converting columns to datetime objects & removing time signatures
autos['date_crawled'] = pd.to_datetime(autos['date_crawled'].str[:10])
autos['ad_created'] = pd.to_datetime(autos['ad_created'])
autos['last_seen'] = pd.to_datetime(autos['last_seen'].str[:10])

autos[['date_crawled','ad_created','last_seen']].head()

date_crawled ad_created  last_seen
0   2016-03-26 2016-03-26 2016-04-06
1   2016-04-04 2016-04-04 2016-04-06
2   2016-03-26 2016-03-26 2016-04-06
3   2016-03-12 2016-03-12 2016-03-15
4   2016-04-01 2016-04-01 2016-04-01

Now to explore the distribution of each column:

In [16]:
print('date_crawled:\n\n',autos['date_crawled'].describe(),'\n')
print('Date          %')
print('----------------------')
print(autos['date_crawled'].value_counts(normalize=True,dropna=False).sort_index())

date_crawled:

 count                   48526
unique                     34
top       2016-04-03 00:00:00
freq                     1873
first     2016-03-05 00:00:00
last      2016-04-07 00:00:00
Name: date_crawled, dtype: object 

Date          %
----------------------
2016-03-05    0.025306
2016-03-06    0.014054
2016-03-07    0.036043
2016-03-08    0.033281
2016-03-09    0.033096
2016-03-10    0.032210
2016-03-11    0.032601
2016-03-12    0.036929
2016-03-13    0.015682
2016-03-14    0.036578
2016-03-15    0.034270
2016-03-16    0.029613
2016-03-17    0.031633
2016-03-18    0.012880
2016-03-19    0.034806
2016-03-20    0.037897
2016-03-21    0.037361
2016-03-22    0.032910
2016-03-23    0.032230
2016-03-24    0.029345
2016-03-25    0.031612
2016-03-26    0.032189
2016-03-27    0.031076
2016-03-28    0.034847
2016-03-29    0.034126
2016-03-30    0.033652
2016-03-31    0.031839
2016-04-01    0.033652
2016-04-02    0.035507
2016-04-03    0.038598
2016-04-04    0.036517
2016-04-05    0.

It looks like the site was crawled over roughly a 1 month period, from March 5th to April 7th of 2016. Other than a small handful of dates, the distribution of listings crawled on each day is mostly uniform.

In [17]:
print('last_seen:\n\n',autos['last_seen'].describe(),'\n')
print('Date          %')
print('----------------------')
print(autos['last_seen'].value_counts(normalize=True,dropna=False).sort_index())

last_seen:

 count                   48526
unique                     34
top       2016-04-06 00:00:00
freq                    10765
first     2016-03-05 00:00:00
last      2016-04-07 00:00:00
Name: last_seen, dtype: object 

Date          %
----------------------
2016-03-05    0.001072
2016-03-06    0.004328
2016-03-07    0.005399
2016-03-08    0.007419
2016-03-09    0.009582
2016-03-10    0.010675
2016-03-11    0.012385
2016-03-12    0.023802
2016-03-13    0.008902
2016-03-14    0.012612
2016-03-15    0.015888
2016-03-16    0.016445
2016-03-17    0.028109
2016-03-18    0.007336
2016-03-19    0.015847
2016-03-20    0.020649
2016-03-21    0.020628
2016-03-22    0.021391
2016-03-23    0.018526
2016-03-24    0.019742
2016-03-25    0.019227
2016-03-26    0.016754
2016-03-27    0.015641
2016-03-28    0.020855
2016-03-29    0.022359
2016-03-30    0.024791
2016-03-31    0.023802
2016-04-01    0.022813
2016-04-02    0.024914
2016-04-03    0.025203
2016-04-04    0.024502
2016-04-05    0.124634

The `last_seen` column encompasses the same time period as the `date_crawled` column. This column helps determine when a listing was removed from the site, presumably because the car was sold. 

The last three days contain a disproportionate amount of `last_seen` values. This likely has more to do with the end of the crawling period than a random spike in car sales.

In [18]:
print('ad_created:\n\n',autos['ad_created'].describe(),'\n')
print('Date          %')
print('----------------------')
print(autos['ad_created'].value_counts(normalize=True,dropna=False).sort_index())

print('\n\nAds Created within the Period of March 5th - April 7th 2016:')
print(autos['ad_created'].value_counts(normalize=True,dropna=False).sort_index()['2016-03-05':'2016-04-07'].sum(),'%')

ad_created:

 count                   48526
unique                     76
top       2016-04-03 00:00:00
freq                     1885
first     2015-06-11 00:00:00
last      2016-04-07 00:00:00
Name: ad_created, dtype: object 

Date          %
----------------------
2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038845
2016-04-04    0.036887
2016-04-05    0.011808
2016-04-06    0.003256
2016-04-07    0.001257
Name: ad_created, Length: 76, dtype: float64


Ads Created within the Period of March 5th - April 7th 2016:
0.9958166755965873 %


The `ad_created` column encompasses a larger period of time, about 10 months between June of 2015 and April of 2016. However, the vast majority listings (over 99%) were created during the same time period as seen in the `date_crawled` & `last_seen` columns.

In [20]:
autos[['registration_year','registration_month']].describe()

registration_year  registration_month
count       48526.000000        48526.000000
mean         2004.754935            5.782694
std            88.678603            3.686184
min          1000.000000            0.000000
25%          1999.000000            3.000000
50%          2004.000000            6.000000
75%          2008.000000            9.000000
max          9999.000000           12.000000

Looking at the final two date related columns, `registration_year` & `registration_month`, everything seems to be in order for the month values, but there are definitely some issues with the year values. 

The `registration_year` column has a minimum value of 1000 and a maximum of 9999. Obviously, both of these values would be impossible for car registration. Certainly, none of the cars in this dataset could have been registered after 2016, and a quick online search shows that Germany did not begin issuing license plates [until the 1900s](https://en.wikipedia.org/wiki/Vehicle_registration_plates_of_Germany#History).

In [38]:
print("Percent of 'registration_year' Values Outside 1900-2016 Range:")
print(100*(~autos['registration_year'].between(1900,2016)).sum()/autos.shape[0],'%')

Percent of 'registration_year' Values Outside 1900-2016 Range:
3.882454766516919 %


Given that the rows outside the 1900-2016 range encompass less than 4% of the dataset, I will remove these rows.

In [44]:
#removing rows outside given registration_year range
autos = autos[autos['registration_year'].between(1900,2016)]
autos['registration_year'].value_counts(normalize=True).head(10)

2000    0.067643
2005    0.062948
1999    0.062111
2004    0.057952
2003    0.057866
2006    0.057245
2001    0.056494
2002    0.053300
1998    0.050598
2007    0.048797
Name: registration_year, dtype: float64

It looks like most of the cars listed were registered in the late 1990s onward. 

## Exploring Price by Brand

In [46]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.211440
bmw               0.110072
opel              0.107671
mercedes_benz     0.096480
audi              0.086617
ford              0.069937
renault           0.047189
peugeot           0.029866
fiat              0.025664
seat              0.018288
skoda             0.016423
nissan            0.015287
mazda             0.015201
smart             0.014172
citroen           0.014022
toyota            0.012714
hyundai           0.010034
sonstige_autos    0.009712
volvo             0.009155
mini              0.008769
mitsubishi        0.008233
honda             0.007847
kia               0.007075
alfa_romeo        0.006646
suzuki            0.005939
chevrolet         0.005703
porsche           0.005574
chrysler          0.003516
dacia             0.002637
daihatsu          0.002508
jeep              0.002273
subaru            0.002144
land_rover        0.002101
saab              0.001651
jaguar            0.001565
daewoo            0.001501
trabant           0.001394
r

The top 5 brands are Volkswagen, BMW, Opel, Mercedes-Benz, and Audi. This is hardly surprising, as these are all German manufacturers. These 5 brands alone account for more than 61% of all listings in the dataset.

Moving forward, I will limit my analysis to common brands, or brands with at least 5% of listings.

In [67]:
common_brands = (
    autos['brand'].value_counts(normalize=True)
    [autos['brand'].value_counts(normalize=True)>.05]
    .index
)
print('Common Brands:')
print(*common_brands,sep='\n')

Common Brands:
volkswagen
bmw
opel
mercedes_benz
audi
ford


In [80]:
#creating dictionary for common brands & mean price values
brand_avg_price = {}

for brand in common_brands:
    avg_price = autos[autos['brand']==brand]['price'].mean()
    brand_avg_price[brand] = avg_price
    
brand_price = pd.DataFrame(pd.Series(brand_avg_price),columns=['Average Listing Price'])
brand_price

Average Listing Price
volkswagen               5402.410262
bmw                      8201.538566
opel                     2975.241935
mercedes_benz            8529.758444
audi                     9295.681683
ford                     3710.766094

The average price of listings for the most common brands hover between about \\$3000 and \\$9300. Both Opel and Ford are on the lower end of this range, while BMW, Mercedes-Benz, and Audi are on the higher end. Volkswagen, the most popular brand, is roughly in the middle.

## Exploring Mileage by Brand

In [82]:
#creating dictionary for common brands & mean odometer_km values
brand_avg_km = {}

for brand in common_brands:
    avg_km = autos[autos['brand']==brand]['odometer_km'].mean()
    brand_avg_km[brand] = avg_km
    
#creating DataFrame w/ mean price & odometer_km values
brand_info = pd.DataFrame(pd.Series(brand_avg_km),columns=['Average Mileage'])
brand_info['Average Price'] = brand_price
brand_info

Average Mileage  Average Price
volkswagen       128707.158791    5402.410262
bmw              132633.424231    8201.538566
opel             129310.035842    2975.241935
mercedes_benz    130851.111111    8529.758444
audi             129188.118812    9295.681683
ford             124288.779890    3710.766094

While there is a *slight* trend toward higher average mileage for the more expensive brands, this trend is pretty negligible. Overall, the average listing price looks to be a much better indicator of brand than average mileage.